test/jdk/java/util/HashMap/WhiteBoxResizeTest.java

```diff
@@ -21,113 +22,290 @@
+        MethodHandles.Lookup hmlookup = MethodHandles.privateLookupIn(HashMap.class, MethodHandles.lookup());
+        TABLE_SIZE_FOR = hmlookup.findStatic(
+            HashMap.class, "tableSizeFor", MethodType.methodType(int.class, int.class));
+        HM_TABLE = hmlookup.unreflectVarHandle(HashMap.class.getDeclaredField("table"));
+
+        MethodHandles.Lookup whmlookup = MethodHandles.privateLookupIn(WeakHashMap.class, MethodHandles.lookup());
+        WHM_TABLE = whmlookup.unreflectVarHandle(WeakHashMap.class.getDeclaredField("table"));
     }
 
+    /*
+     * utility methods
+     */
+
     int tableSizeFor(int n) {
         try {
             return (int) TABLE_SIZE_FOR.invoke(n);
         } catch (Throwable t) { throw new AssertionError(t); }
     }
 
-    Object[] table(HashMap map) {
+    Object[] table(Map<?,?> map) {
         try {
-            return (Object[]) TABLE.get(map);
+            VarHandle vh = map instanceof WeakHashMap ? WHM_TABLE : HM_TABLE;
+            return (Object[]) vh.get(map);
         } catch (Throwable t) { throw new AssertionError(t); }
     }
 
-    int capacity(HashMap map) {
+    int capacity(Map<?,?> map) {
         return table(map).length;
     }
 
-    @Test
-    public void testTableSizeFor() {
-        assertEquals(tableSizeFor(0), 1);
-        assertEquals(tableSizeFor(1), 1);
-        assertEquals(tableSizeFor(2), 2);
-        assertEquals(tableSizeFor(3), 4);
-        assertEquals(tableSizeFor(15), 16);
-        assertEquals(tableSizeFor(16), 16);
-        assertEquals(tableSizeFor(17), 32);
-        int maxSize = 1 << 30;
-        assertEquals(tableSizeFor(maxSize - 1), maxSize);
-        assertEquals(tableSizeFor(maxSize), maxSize);
-        assertEquals(tableSizeFor(maxSize + 1), maxSize);
-        assertEquals(tableSizeFor(Integer.MAX_VALUE), maxSize);
+    // creates a map with size mappings
+    Map<Integer, Integer> makeMap(int size) {
+        return IntStream.range(0, size)
+                        .boxed()
+                        .collect(Collectors.toUnmodifiableMap(i -> i, i -> i));
+    }
+
+    // creates a "fake" map: size() returns the given size, but
+    // the entrySet iterator returns only one entry
+    Map<Integer, Integer> fakeMap(int size) {
+        return new AbstractMap<>() {
+            public Set<Map.Entry<Integer,Integer>> entrySet() {
+                return new AbstractSet<Map.Entry<Integer,Integer>>() {
+                    public int size() { return size; }
+                    public Iterator<Map.Entry<Integer,Integer>> iterator() {
+                        return Set.of(Map.entry(1, 1)).iterator();
+                    }
+                };
+            }
+        };
     }
 
-    @Test
-    public void capacityTestDefaultConstructor() {
-        capacityTestDefaultConstructor(new HashMap<>());
-        capacityTestDefaultConstructor(new LinkedHashMap<>());
+    void putN(Map<Integer, Integer> map, int n) {
```

Comment:
@stuart-marks well we know this is correct for WeakHashMap when n < IntegerCache.high because we have Integer constant pool, but I think it be better to manlly make it sure it is referenced, as IntegerCache.high is configurable.

Reply:
@stuart-marks please have a look in changes in the latest commit, I think we'd better to manually create references for keys like that.

Reply:
Good point about WeakHashMap! I don't think we need a separate table. Since the value is held by a strong reference, using the same reference as the key will prevent the key's weak reference from being cleared. I'd suggest this:

    for (int i = 0; i < n; i++) {
        Integer obj = i; // prevent WeakHashMap from clearing keys
        map.put(obj, obj);
    }

Reply:
Heads up here!
java.lang.Integer is specified  as a value based class and should not be used where identity is needed.
https://docs.oracle.com/en/java/javase/17/docs/api/java.base/java/lang/doc-files/ValueBased.html

I don't have a ready suggestion, but be forewarned that this use is not considered valid.

Reply:
> Heads up here! java.lang.Integer is specified as a value based class and should not be used where identity is needed. https://docs.oracle.com/en/java/javase/17/docs/api/java.base/java/lang/doc-files/ValueBased.html
> 
> I don't have a ready suggestion, but be forewarned that this use is not considered valid.


@RogerRiggs

Sounds reasonable.

Could you please have a look at the latest commit? I tried to make a class to use as a key, instead of using integer.